libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp

```diff
@@ -20,49 +21,135 @@ void AP_BattMonitor_SMBus::init(void)
     }
 }
 
-// return true if cycle count can be provided and fills in cycles argument
+// return true if cycle count has been read and fills in cycles argument
 bool AP_BattMonitor_SMBus::get_cycle_count(uint16_t &cycles) const
 {
-    if (!_has_cycle_count) {
+    WITH_SEMAPHORE(_dev->get_semaphore());
```

Comment:
This is not a good idea. We should be reading this in the driver and stashing if we want it. The problem with exposing this is we start doing transactions from other threads, and since we configure the bus as SMBus this actually has a *very* long timeout on the transaction. If the battery (or something else) is very busy in the timer waiting on the hardware to catch up this could cause the main thread to stall while fetching the information for MAVLink for example.

Reply:
I understand your concerns and will revert those changes. 

The reason I was attempting this method was to limit storage and simplify functions required for rarely accessed info. Once the info. gets sent via mavlink or logged it no longer needs to be stored. I'm not sure if there is a way to accomplish that while limiting delays/timeouts to the battmon thread? 

Reply:
You can arbitrarily request a message from MAVLink though, and you might need to ensure someone who connects late gets it, which would argue for keeping it. The easiest way to minimize it is to store all the needed information in a struct you only allocated if any of the needed data was available.

Reply:
For the arbitrary mavlink request or late connection I was thinking of setting a flag to signal the battmon timer loop to handle building the message. (But I couldn't figure out how to do that. For logging it was obvious and I got that working in that manner.)

Reply:
set back to stashing